vncviewer/touch.cxx

```diff
@@ -47,12 +54,39 @@ static rfb::LogWriter vlog("Touch");
 
 #if !defined(WIN32) && !defined(__APPLE__)
 static int xi_major;
+#endif
 
-typedef std::map<Window, class XInputTouchHandler*> HandlerMap;
+typedef std::map<Window, class BaseTouchHandler*> HandlerMap;
 static HandlerMap handlers;
-#endif
 
-#if !defined(WIN32) && !defined(__APPLE__)
+#if defined(WIN32)
+LRESULT CALLBACK win32WindowProc(HWND hWnd, UINT uMsg, WPARAM wParam,
+                                 LPARAM lParam, UINT_PTR uIdSubclass,
+                                 DWORD_PTR dwRefData)
+{
+  bool handled = false;
+
+  if (uMsg == WM_NCDESTROY) {
+    delete handlers[hWnd];
+    handlers.erase(hWnd);
+    RemoveWindowSubclass(hWnd, &win32WindowProc, 1);
+
+  } else {
+    if (handlers.count(hWnd) == 0)
+      vlog.error(_("Got message (0x%x) for an unhandled window"), uMsg);
+    else
+      handled = dynamic_cast<Win32TouchHandler*>
+        (handlers[hWnd])->processEvent(uMsg, wParam, lParam);
+  }
+
+  // Only run the normal WndProc handlers for unhandled events
+  if (handled)
+    return 1;
```

Comment:
How do we know this is the correct return value?

Reply:
Apparently WM_GESTURENOTIFY always want handlers to run DefWindowProc, and WM_GESTURE should have a return value of `0` when handled.

I changed processEvent for WM_GESTURENOTIFY to mark it as unhandled, and modified the return for "handled" cases.